### PR TITLE
pool/SNOW-902611 #1/4 removed singleton pattern from SessionPool; introduced ne…

### PR DESCRIPTION
**Pull request by sfc-gh-mhofman**
### Description
Connection Pool v1
-Removal of singleton pattern from SessionPool
-Introduction of interface and an its implementation of ConnectionManager
-Split integration tests of ConnectionPool (into separate files) + some cleanup in the classes

### Checklist
- [x] Code compiles correctly
- [x] Code is formatted according to [Coding Conventions](../CodingConventions.md)
- [x] Created tests which fail without the change (if possible)
- [x] All tests passing (`dotnet test`)
- [x] Extended the README / documentation, if necessary
- [x] Provide JIRA issue id (if possible) or GitHub issue id in PR name

### CODE REVIEW

File: Snowflake.Data.Tests/IntegrationTests/SFConnectionPoolAsyncIT.cs [hunk 1]
```diff
@@ -0,0 +1,361 @@
...
+    using System.Threading;
+    using System.Threading.Tasks;
+    using NUnit.Framework;
+    using Moq;
+    using Snowflake.Data.Client;
+    using Snowflake.Data.Core;
+    using Snowflake.Data.Tests.Mock;
+
+    [TestFixture, NonParallelizable]
+    class SFConnectionPoolITAsync : SFBaseTestAsync
```
Comment (sfc-gh-mhofman) [inline Snowflake.Data.Tests/IntegrationTests/SFConnectionPoolAsyncIT.cs:21]: Extracted class from file SFConnectionPoolT. Minor changes in couple of lines to remove IDE highlights

File: Snowflake.Data/Core/Session/IConnectionManager.cs [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+ * Copyright (c) 2012-2023 Snowflake Computing Inc. All rights reserved.
+ */
+
+using System.Security;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Snowflake.Data.Core.Session
+{
+    internal interface IConnectionManager
```
[the remark this answers is not in the document]
Comment (sfc-gh-mhofman) [reply Snowflake.Data/Core/Session/IConnectionManager.cs:11]: I would say more about connection management as the implementation relates to handling multiple connection pools at a time. However, connection and session are strongly linked

File: Snowflake.Data/Core/Session/SessionPool.cs [hunk 3]
```diff
@@ -9,20 +13,18 @@
 
 namespace Snowflake.Data.Core.Session
 {
-    sealed class SessionPoolSingleton : IDisposable
+    sealed class SessionPool : IDisposable
```
Comment (sfc-gh-dstempniak) [inline Snowflake.Data/Core/Session/SessionPool.cs:16]: Maybe call it SessionCache?

File: Snowflake.Data/Core/Session/SessionPool.cs [hunk 4]
```diff
@@ -9,20 +13,18 @@
...
     {
-        private static readonly SFLogger s_logger = SFLoggerFactory.GetLogger<SessionPoolSingleton>();
-        private static SessionPoolSingleton s_instance = null;
+        private static readonly SFLogger s_logger = SFLoggerFactory.GetLogger<SessionPool>();
         private static readonly object s_sessionPoolLock = new object();
-
         private readonly List<SFSession> _sessionPool;
```
[the remark this answers is not in the document]
Comment (sfc-gh-dstempniak) [reply Snowflake.Data/Core/Session/SessionPool.cs:20]: Or _sessionCache ...

File: Snowflake.Data/Core/Session/ConnectionManagerV1.cs [hunk 5]
```diff
@@ -0,0 +1,23 @@
+using System.Security;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Snowflake.Data.Core.Session
+{
+    internal sealed class ConnectionManagerV1 : IConnectionManager
```
Comment (sfc-gh-dstempniak) [inline Snowflake.Data/Core/Session/ConnectionManagerV1.cs:7]: ConnectionCacheManager ?

File: Snowflake.Data.Tests/IntegrationTests/SFConnectionPoolIT.cs [hunk 6]
```diff
@@ -0,0 +1,411 @@
...
+    using System.Threading;
+    using System.Threading.Tasks;
+    using NUnit.Framework;
+    using Snowflake.Data.Core;
+    using Snowflake.Data.Client;
+
+    [TestFixture, NonParallelizable]
+    class SFConnectionPoolIT : SFBaseTest
+    {
+        private static readonly PoolConfig s_previousPoolConfig = new PoolConfig();
```
[the remark this answers is not in the document]
Comment (sfc-gh-mhofman) [reply Snowflake.Data.Tests/IntegrationTests/SFConnectionPoolIT.cs]: done

File: Snowflake.Data.Tests/IntegrationTests/SFConnectionPoolAsyncIT.cs [hunk 7]
```diff
@@ -0,0 +1,361 @@
...
+        [Test]
+        public void TestConnectionPoolWithAsync()
+        {
+            using (var conn = new MockSnowflakeDbConnection())
+            {
+                SnowflakeDbConnectionPool.SetMaxPoolSize(1);
+                SnowflakeDbConnectionPool.ClearAllPools();
+
+                int timeoutSec = 0;
+                string infiniteLoginTimeOut = String.Format("" + ";connection_timeout={0}",
```
[the remark this answers is not in the document]
Comment (sfc-gh-mhofman) [reply Snowflake.Data.Tests/IntegrationTests/SFConnectionPoolAsyncIT.cs]: done

File: Snowflake.Data.Tests/IntegrationTests/SFConnectionPoolAsyncIT.cs [hunk 8]
```diff
@@ -0,0 +1,361 @@
...
+                Assert.AreEqual(conn.State, ConnectionState.Closed);
+
+                CancellationTokenSource connectionCancelToken = new CancellationTokenSource();
+                try
+                {
+                    conn.OpenAsync(connectionCancelToken.Token);
+                }
+                catch (SnowflakeDbException ex)
+                {
+                    Console.WriteLine("connection failed:" + ex);
```
[the remark this answers is not in the document]
Comment (sfc-gh-mhofman) [reply Snowflake.Data.Tests/IntegrationTests/SFConnectionPoolAsyncIT.cs]: replaced with logger

File: Snowflake.Data/Core/Session/SessionPool.cs (hunk 3 above)
Comment (sfc-gh-mhofman) [reply Snowflake.Data/Core/Session/SessionPool.cs:16]: this name was for long here and it is actually a mechanism to pool sessions so why to change?

File: Snowflake.Data/Core/Session/SessionPool.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (sfc-gh-mhofman) [reply Snowflake.Data/Core/Session/SessionPool.cs:20]: _sessions is better, will change

[the remark this answers is not in the document]
Comment (sfc-gh-mhofman) [reply Snowflake.Data/Core/Session/SessionPool.cs:20]: will do

**Review by sfc-gh-knozderko — APPROVED** [review]

**Review by sfc-gh-dstempniak — APPROVED** [review]

**Pull request merged**